src/ray/object_manager/object_buffer_pool.cc

```diff
@@ -117,24 +117,28 @@ std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> ObjectBufferPool::Cr
         plasma::flatbuf::ObjectSource::ReceivedFromRemoteRaylet);
     lock.lock();
 
-    std::vector<boost::asio::mutable_buffer> buffer;
-    if (!s.ok()) {
-      // Create failed. The object may already exist locally. If something else went
-      // wrong, another chunk will succeed in creating the buffer, and this
-      // chunk will eventually make it here via pull requests.
-      return std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status>(
-          errored_chunk_, ray::Status::IOError(s.message()));
+    // Another thread may have succeeded in creating the chunk while the lock
+    // was released. In that case skip the remainder of the creation block.
+    if (create_buffer_state_.count(object_id) == 0) {
+      std::vector<boost::asio::mutable_buffer> buffer;
+      if (!s.ok()) {
+        // Create failed. The object may already exist locally. If something else went
+        // wrong, another chunk will succeed in creating the buffer, and this
+        // chunk will eventually make it here via pull requests.
+        return std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status>(
+            errored_chunk_, ray::Status::IOError(s.message()));
+      }
+      // Read object into store.
+      uint8_t *mutable_data = data->Data();
```

Comment:
Shouldn't the copy take place outside of the `if` block still?

Reply:
Yeah good point. Let me look into this more, I think this function is quite brittle and we could rewrite it to be more resilient to duplicate chunk puts.

Reply:
Turns out this was only to initialize the buffer (the actual copy is done in ObjectManager).